### ISSUE TITLE
[Question] How to append online transitions with pre-existing d4rl buffer in finetuning training?

### ISSUE THREAD

**Issue by HYDesmondLiu** [takuseno/d3rlpy#279]
If I have pre-trained offline models and I have loaded them as policy.
Then I want to append new transitions into existing d4rl buffers. How do I do it?

In the documentation and tutorial, there is only instruction about creating a new buffer for finetuning. 
And since `buffer` and `dataset` are not the same datatype, how do I append new transitions and discard `buffer`

I initialized the `buffer`, and then append transitions from the `dataset`. 
However, if then I use `policy.fit_online(env, buffer, n_steps=eval_freq)` then the buffer is not adding new online transitions. (where the `policy` is pre-trained `CQL`).

Even if I use `d3rlpy.online.buffers.ReplayBuffer.append(observations, action, reward, terminal)`, or `policy.collect(env, buffer)`
The new transitions are not appended to the buffer.

Here is how I initialize the model and load it with the pre-trained offline model:

```policy = d3rlpy.algos.CQL(use_gpu=True)
dataset, env = d3rlpy.datasets.get_dataset(env_name)
policy.build_with_dataset(dataset) 
policy.load_model(pretrained_model_name)
buffer = d3rlpy.online.buffers.ReplayBuffer(maxlen=1000000, env=env)   
```

**Comment by takuseno** [comment]
@HYDesmondLiu Hi, thanks for the issue. Please check this IQL reproduction script.
https://github.com/takuseno/d3rlpy/blob/a11b15db1e60d47296fdc7e7880fe1165803138a/reproductions/finetuning/iql.py#L61

**Comment by HYDesmondLiu** [comment]
Thanks for the prompt reply.
OKAY, let me ask you this way. Since the buffer is FIFO, I have tried to fill the `buffer` with the dataset's transitions.
And with `CQL.fit_online`, the first transition is not changed after finetuning via interacting with the environment. 


My workaround is to enlarge the `maxlen` of the buffer, I believe there is a bug in the `append` code block.

**Comment by takuseno** [comment]
Can you make sure if you set `episodes=dataset.episodes` like the example above?

**Comment by HYDesmondLiu** [comment]
I have set it and it does not work either. 
Since it is a queue, the first element should be removed when we try to add a new element to the full buffer.
However, I use `buffer.transitions[0]` to print out the first observation in the buffer. It is always the same. 

Basically, I was doing the training as the following snippet. Please let me know if anything is set incorrectly that 
leads to the buffer is not updated after training online.
````
dataset, env = d3rlpy.datasets.get_dataset(env_name)
policy = d3rlpy.algos.CQL(use_gpu=True)
pretrained_model_name = f'xxx.pt'
policy.build_with_env(env)
policy.load_model(pretrained_model_name)
buffer = d3rlpy.online.buffers.ReplayBuffer(maxlen=1_000_000, env=env, episodes=dataset.episodes) 
policy.fit_online(env, buffer, n_steps=250_000)
t0=buffer.transitions[0]
tl=buffer.transitions[-1]
print(f'first obs.:{t0.observation}/last obs.:{tl.observation}')
````

**Comment by takuseno** [comment]
Which dataset are you using? Because let's say the maximum capacity of the replay buffer is 1M and the dataset size is below 1M. In that case, there is still room to store new experiences replay buffer. If `n_steps` of `fit_online` is very large such as 1M, do you still see the same result?

**Comment by HYDesmondLiu** [comment]
I am using MuJoCo, however, I deliberately downsample the dataset on purpose. 

Since in real-world offline-to-online finetuning problems there might not be access to while offline dataset. 

So actually I downsample the buffer with only 5% of the original amount. So the buffers are filled up really quick.

**Comment by takuseno** [comment]
I see. Then, I believe that the replay buffer still has space to store new experiences. You can test this by decreasing `maxlen` of the replay buffer to the very small value to see if it works correctly.

**Comment by HYDesmondLiu** [comment]
Thanks for quick response. Actually that was what I did. The code snippet I provided is modified. The `maxlen` I used was only 5% of the original size of the buffer. And I output the first element of the observations after every `fit_online` was done. And it was always the same one.

**Comment by takuseno** [comment]
I couldn't actually reproduce your problem on my end. Here is my test code:
```
import d3rlpy

dqn = d3rlpy.algos.DQN(batch_size=10)
dataset, env = d3rlpy.datasets.get_cartpole()
buffer = d3rlpy.online.buffers.ReplayBuffer(maxlen=100, env=env, episodes=dataset.episodes)
print(buffer.transitions[0].observation)
dqn.fit_online(env, buffer, n_steps=100)
print(buffer.transitions[0].observation)
```
Note that I'm using `gym==0.17.0` for compatibility issues, but this does not affect the behavior of `ReplayBuffer`. The result is:
```
[0.37582362 0.5620217  0.00999125 0.24562645]
[-0.00132633 -0.599253   -0.00662717  0.88046914]
```

**Comment by HYDesmondLiu** [comment]
Thanks for the instruction and replies.
I have created a snippet as a simplified example. And now it works. 
There must be something I have done wrong in previous codes. 

```
import d3rlpy

dataset, env = d3rlpy.datasets.get_dataset("hopper-expert-v2")
policy = d3rlpy.algos.CQL(use_gpu=True)
policy.fit(dataset, n_steps = 100, n_steps_per_epoch=50, show_progress=False)
policy.save_model(f'CQL.pt')

pretrained_model_name = f'CQL.pt'
policy.build_with_env(env)
policy.load_model(pretrained_model_name)


buffer = d3rlpy.online.buffers.ReplayBuffer(maxlen=20, env=env, episodes=dataset.episodes)    

for i in range(2):
    policy.fit_online(env, buffer, n_steps=100, n_steps_per_epoch=50, show_progress=False)
    t0=buffer.transitions[0]
    print(f'first obs.:{t0.observation}')
  
```
